tests/ballerina-integration-test/src/test/java/org/ballerinalang/test/service/websub/WebSubServiceExtensionTestCase.java

```diff
@@ -0,0 +1,169 @@
+    private static final String BY_HEADER_AND_PAYLOAD_FEATURE_PULL_LOG =
+            "Feature Pull Notification Received, header value: pull domain: feature";
+    private static final String BY_HEADER_AND_PAYLOAD_HEADER_ONLY_LOG =
+            "HeaderOnly Notification Received, header value: headeronly action: header_only";
+    private static final String BY_HEADER_AND_PAYLOAD_KEY_ONLY_LOG =
+            "KeyOnly Notification Received, header value: key_only action: keyonly";
+
+    private LogLeecher byKeyCreatedLogLeecher = new LogLeecher(BY_KEY_CREATED_LOG);
+    private LogLeecher byKeyFeatureLogLeecher = new LogLeecher(BY_KEY_FEATURE_LOG);
+
+    private LogLeecher byHeaderIssueLogLeecher = new LogLeecher(BY_HEADER_ISSUE_LOG);
+    private LogLeecher byHeaderCommitLogLeecher = new LogLeecher(BY_HEADER_COMMIT_LOG);
+
+    private LogLeecher byHeaderAndPayloadIssueCreatedLogLeecher =
+            new LogLeecher(BY_HEADER_AND_PAYLOAD_ISSUE_CREATED_LOG);
+    private LogLeecher byHeaderAndPayloadFeaturePullLogLeecher = new LogLeecher(BY_HEADER_AND_PAYLOAD_FEATURE_PULL_LOG);
+    private LogLeecher byHeaderAndPayloadHeaderOnlyLogLeecher = new LogLeecher(BY_HEADER_AND_PAYLOAD_HEADER_ONLY_LOG);
+    private LogLeecher byHeaderAndPayloadKeyOnlyLogLeecher = new LogLeecher(BY_HEADER_AND_PAYLOAD_KEY_ONLY_LOG);
+
+    @BeforeClass
+    public void setup() throws BallerinaTestException {
+        String subscriberBal = new File("src" + File.separator + "test" + File.separator + "resources"
+                                                + File.separator + "websub" + File.separator
+                                                + "websub_custom_subscriber_service.bal").getAbsolutePath();
+
+        webSubSubscriber.addLogLeecher(byKeyCreatedLogLeecher);
+        webSubSubscriber.addLogLeecher(byKeyFeatureLogLeecher);
+        webSubSubscriber.addLogLeecher(byHeaderIssueLogLeecher);
+        webSubSubscriber.addLogLeecher(byHeaderCommitLogLeecher);
+        webSubSubscriber.addLogLeecher(byHeaderAndPayloadIssueCreatedLogLeecher);
+        webSubSubscriber.addLogLeecher(byHeaderAndPayloadHeaderOnlyLogLeecher);
+        webSubSubscriber.addLogLeecher(byHeaderAndPayloadFeaturePullLogLeecher);
+        webSubSubscriber.addLogLeecher(byHeaderAndPayloadKeyOnlyLogLeecher);
+
+        webSubSubscriber.startBallerinaServer(subscriberBal, new String[]{}, 8181);
+    }
+
+    @Test
+    public void testDispatchingByKey() throws BallerinaTestException, IOException {
+        Map<String, String> headers = new HashMap<>();
+        headers.put(HttpHeaderNames.CONTENT_TYPE.toString(), TestConstant.CONTENT_TYPE_JSON);
+        HttpResponse response = HttpClientRequest.doPost("http://localhost:8181/key", "{\"action\":\"created\"}",
+                                                         headers);
+        Assert.assertEquals(response.getResponseCode(), 202);
+        byKeyCreatedLogLeecher.waitForText(3000);
```

Comment:
Could have added a constant for the response code/timeout as these are used in multiple places